torchtitan/checkpoint.py

```diff
@@ -137,6 +137,38 @@ def __init__(
 
         if not self.enable_checkpoint:
             return
+        """
+        Note: Pipeline Parallelism and Virtual Stages
+
+        1. even for simple PP schedules, there is a separate optimizer each PP rank.
+        rank0's optimizer would have a param_group[0] which refers to layers.0 in the original model.
+        rank1's would _also_ have a param_group[0], since it's index based, but referring to layers.1.
+        When saving, these collide and one of them is lost.  Then when reloading, only one stage can
+        restore its optimizer states, others will error.
+
+            The solution to this problem is optimizer flattening: it landed in #127071 and is enabled in TorchTitan
+            by passing the 'flatten_optimizer_state_dict' kwarg to DCP functions called in the OptimizerWrapper.
+
+        2. With complex PP schedules, we have multiple model chunks per pp rank. This compounds challenge (1) by also
+        requiring us to reason about multiple 'optim' objects locally.
+
+            We solve this in the Model and Optimizer wrapper classes by flattening the state dicts from each object
+            into one state dict before saving/loading. We rely on the individual state_dicts to not collide,
+            which is gauranteed for the model by correct pipeline splitting and for the optimizer by the flattening
+            support described in (1).
+
+        3. LR schedulers also index model states like optimizers and would need to be flattened properly to support
```

Comment:
RE:
<img width="967" alt="image" src="https://github.com/pytorch/torchtitan/assets/4984825/21264a52-80c5-4019-987a-ce0e6d17bfa6">

I did it this way because i thought if we are not supporting resharding of lr_scheduler, then i may as well save each one.  If i save each one, then at load time i have a form of assertion provided for me by dcp- if the runtime loading the checkpoint has the same number of ranks, they will match up and load OK. if not, they would throw an error.

I could switch back to the version where I save only one copy, but then i have to do some validation up front.  Do you think this way is better?
@fegin

Reply:
If we are not supporting resharding, then this implementation is better.